### ISSUE TITLE
Nested track_caller

### ISSUE THREAD

**Issue by orenbenkiki** [rust-lang/rust-clippy#9472]
### What it does

Given a top-level function annotated by `track_caller`, which invokes intermediate function(s) without `track_caller`, which invoke low-level function(s) with `track_caller`, then warn that the intermediate functions should probably have `track_caller` annotation as well.

### Lint Name

intermediate_track_caller

### Category

suspicious

### Advantage

The above scenario is likely a bug.

When a top-level function specifies `track_caller` it indicates that internal panics should be reported at the caller's site. This is easy to manually annotate when defining the function's interface, especially when adding a `# Panics` section in the documentation.

A low-level assertion-like function which panics on some condition is also easy to annotate with `track_caller`. Still, clippy may help with this, e.g. #6835.

However, it is very easy to omit the `track_caller` annotation for intermediate logic functions. Paradoxically this becomes more difficult the more neatly the code base is nicely factored out to assertion-like functions, internal logic functions, and top-level API functions.

Having clippy track this issue would ensure that all code paths from a top-level function to the low-level assertion-like functions are annotated with `track_caller` so that panics would be reported in the invocation of the top-level function rather than in the internal intermediate logic functions.

### Drawbacks

The main disadvantage (other than implementation complexity) is the risk for false positives.

In general, there are two kind of panics a library or a module may generate - panics indicating the user has broken the API contract and panics that indicate an internal bug in the code. Ideally, "user panics" would be reported on the invocation of a top-level API function and "internal panics" would be reported on the panicking line.

The `track_caller` mechanism isn't friendly for this distinction; it assumes that every panic call in the annotated function is a "user panic" and any panic call in a nested function is an "internal panic".

The provided lint assumes in contrast that every use of `track_caller` in the crate indicates a "user panic" and tries to ensure that by having all the code paths leading to it annotated by `track_caller` as well. This is a sensible assumption but it does not always hold.

Specifically, if a crate contains two (or more) logically separate layers, each may have its own API and internal logic and its own two kinds of panics, in which case the lint will not apply. Arguably it would be better to split such a crate into two sub-crates, one for each layer, to get the advantage of the lint in each one. However this may not be practical in every scenario. In such cases it would be better to disable the lint for the crate.

### Example

```rust
#[track_caller]
pub fn foo(x: i32) {
    bar(x);
}

fn bar(x: i32) {
   baz(x);
}

#[track_caller]
fn baz(x: i32) {
    assert!(x > 0);
}
```

Could be written as:

```rust
#[track_caller]
pub fn foo(x: i32) {
    bar(x);
}

#[track_caller] // <--- Added
fn bar(x: i32) {
   baz(x);
}

#[track_caller]
fn baz(x: i32) {
    assert!(x > 0);
}
```
